### PR TITLE
Zapp Brannigan: Convert to full template

### LINKED ISSUE #2767: Convert Simple Structured Answer Goodies to Full Templates
### Background

Goodies have full support for templates, and we'd like to make sure all Goodies are using our full templates. We provided the "simple structured_answer" as a way to get them using basic templates before they had full template support and now the time has come to upgrade them all.

### What To Do

1. Change the returned `structured_answer` object (in one of the .pm files below) to match the new format, which specifies a template and the data
    
    Nearly all of these Goodies should be able to use our simple `text` template and should return a `structured_answer` that looks like the one, returned by the PigLatin Goodie in [PigLatin.pm](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/PigLatin.pm#L40).

2. Update the Instant Answer's test file to match the changes made. The tests verify the expected content of the `structured_answer` for a given query. 

3. Test the IA locally with DuckPAN to make sure it still looks good and the tests pass. 

    You can run the test file with `duckpan test`:
    Example:
    Anagram.pm => `duckpan test Anagram`
    AspectRatio.pm => `duckpan test AspectRatio`

    You can test the visual display using `duckpan server`. You'll want to try some of the queries in the test file and make sure the visuals look good. Ideally identical to how they currently look on the live site.
    
### Example

@GuiltyDolphin has made [a great Pull Request](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2769/files) that demonstrates the necessary changes for the Anagram Instant Answer:

anagrams__update_to_full_template_by_guiltydolphin_ _pull_request__2769_ _duckduckgo_zeroclickinfo-goodies

When updating the test file, it will make things _**much**_ easier if you create a small function that generates a structured answer object. This way you won't need to manually type out the full structured answer we expect for each individual query.

@GuiltyDolphin has done this in the Anagram PR as well:
anagrams__update_to_full_template_by_guiltydolphin_ _pull_request__2769_ _duckduckgo_zeroclickinfo-goodies

anagrams__update_to_full_template_by_guiltydolphin_ _pull_request__2769_ _duckduckgo_zeroclickinfo-goodies

#### Note

Some of these IAs may actually be offline right now and so you won't be able to trigger them on DuckDuckGo.com. Once you've made a Pull Request we will deploy it to beta.duckduckgo.com so you can view your changes in a production environment

### How To Get Started

1. Find an Instant Answer in the list below that doesn't have a name beside it.
2. Leave a comment below that indicates which IA(s) you'd like to tackle. One of the community leaders or staff will put your name beside the IA in the list below indicating that it's been claimed.
3. Get coding and make a Pull Request.
    Be sure to include the link to the related Instant Answer Page in your PR Comment. This is **required** for us to be able to deploy your PR to Beta and ultimately merge it. For example, the anagram PR links to https://duck.co/ia/view/anagram.

    Also, make sure you mention this issue in your PR. That way we can keep track of all the PR's in here as well. Something like "Updates Anagram Goodie Re: #2767" is good.

----

Here's the list of IA's that need to be converted. If you'd like to work on one, please leave a comment letting everyone know, so we can avoid and duplicated efforts.

- [x] Anagram.pm        - @GuiltyDolphin
- [x] AspectRatio.pm    - @Mylainos
- [x] Base.pm    - @Mylainos
- [x] Average.pm   - @Mylainos
- [x] Base64.pm     - @Mylainos
- [x] BeamMeUpScotty.pm
- [x] Binary.pm   - @Mylainos
- [x] BirthStone.pm - @rasikapohankar
- [x] Bitsum.pm   - @Mylainos
- [x] Braille.pm -  https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2811
- [x] BRT.pm   - @Mylainos
- [x] CalcRoots.pm       - @GuiltyDolphin
- [x] CalendarConversion.pm   - @Mylainos
- [x] Calculator.pm        - @GuiltyDolphin
- [x] CallingCodes.pm   - @Mylainos
- [x] Chars.pm
- [x] CoffeeToWaterRatio.pm - @lamanh
- [x] Coin.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3128
- [x] Combination.pm
- [x] Constants.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3051
- [x] ConvertLatLon.pm - @mintsoft
- [x] CountryCodes.pm - @NaveenKarippai
- [x] CurrencyIn.pm
- [x] DaysBetween.pm - @NaveenKarippai https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3117
- [x] Factors.pm - @mintsoft
- [x] EmailValidator.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3163
- [x] Fibonacci.pm - @mintsoft
- [x] FlipText.pm - @NaveenKarippai
- [x] Fortune.pm - @mintsoft
- [x] GenerateMAC.pm - @mintsoft
- [x] GreatestCommonFactor.pm
- [x] GUID.pm - @gaulrobe #3007
- [x] HelpLine.pm - @mintsoft
- [x] HexToASCII.pm - @dbech
- [x] HexToDec.pm - @mintsoft
- [x] HTMLEntitiesDecode.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3041
- [x] ISO639.pm - @mintsoft
- [x] Jira.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3069
- [x] Kana.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3072
- [x] LeetSpeak.pm - @mintsoft
- [x] Lowercase.pm
- [x] MacAddress.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3076
- [x] MagicEightBall.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3077
- [x] MakeMeASandwich.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3078
- [x] MD4.pm - @mintsoft
- [x] MD5.pm - @mintsoft
- [x] MoonPhases.pm
- [x] Morse.pm
- [x] NoteFrequency.pm
- [x] PaleoIngredientCheck.pm
- [x] Paper.pm - @mintsoft
- [x] ParseCron.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3125
- [x] Passphrase.pm  - @mintsoft
- [x] Password.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3132
- [x] PercentOf.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3131
- [x] PhoneAlphabet.pm - @mintsoft
- [x] Pi.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3130
- [x] POTUS.pm
- [x] Rafl.pm - @mintsoft
- [x] RandomNumber.pm - @NaveenKarippai
- [x] Rc4.pm- @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3098
- [x] Reverse.pm - @mintsoft
- [x] ReverseComplement.pm - @mintsoft
- [x] RIPEMD.pm - @mintsoft
- [x] Roman.pm - @mintsoft
- [x] Rot13.pm - @mintsoft
- [x] Scramble.pm - @mintsoft
- [x] Sha.pm - @gaulrobe https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3119
- [x] SHA3.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3122
- [X] SolarSystem.pm
- [x] Sort.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3123
- [x] BeamMeUpScotty.pm
- [x] TimezoneConverter.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3124
- [x] TitleCase.pm
- [x] TwelveOclock.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3127
- [x] Uppercase.pm
- [x] Wavelength.pm - @mintsoft
- [x] Week.pm  - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3126
- [x] WeekdaysBetween.pm - https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3104
- [x] Weight.pm
- [x] ValarMorghulis.pm
- [x] WhereAmI.pm - https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2684
- [x] WorkdaysBetween.pm
- [x] ZappBrannigan.pm - @rasikapohankar

---

/cc @duckduckgo/community-leaders

### PR DESCRIPTION
**Pull request by rasikapohankar**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [ ] New Instant Answer
    - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
    - [ ] Other: **`New {IA Name} Instant Answer`**
- [ ] Improvement
    - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
    - [x] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{GoodieRole/Templates/Tests/Docs}: {Short Description}`**

###### Description of changes

Provide an overview of the changes this pull request introduces.

###### Which issues (if any) does this fix?

Fixes #2767
Updates ZappBrannigan Goodie to full template. Added sub-template for displaying new lines in the quote.

---
People to notify (@mention interested parties)
@GuiltyDolphin @moollaza 

Instant Answer Page: https://duck.co/ia/view/zapp_brannigan

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @majuscule

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by majuscule** [comment]
lgtm. maybe add some italics to the peoples names while we're at it so it's a bit easier to read?

**Comment by rasikapohankar** [comment]
@majuscule I'll try it. Would we do that by adding an `<i>` tag around the names in the quotes.yml file?

**Comment by majuscule** [comment]
That sounds fine to me. It'd be better to get a designer looking at it but it doesn't really matter. I don't use this goodie personally, so I'd be happy to have it merged as is with no aesthetic changes. Not sure what the workflow is here as I am marked the assignee and maintainer but do not have merge rights. @moollaza

**Comment by rasikapohankar** [comment]
@majuscule As far as I know the maintainer goes through whatever changes are made and once they are happy with it and everything works, someone with access will merge the changes.

**Comment by rasikapohankar** [comment]
@GuiltyDolphin @mintsoft In the pull request there are 3 changed files but there are 4 commits, two are old. Can you please tell me if this will be a problem? There is no Travis build success or failure mark on the new commits. What did I do wrong?

File: lib/DDG/Goodie/ZappBrannigan.pm [hunk 1]
```diff
@@ -19,12 +19,20 @@ handle query => sub {
     # Ensure rand is seeded for each process
     srand();
     my @quote = @{$quotes->[int(rand(scalar(@$quotes)))]};
-
+
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ZappBrannigan.pm]: Lets clean that up!

**Comment by GuiltyDolphin** [comment]
@rasikapohankar It shouldn't be an issue, it looks like they are commits that are part of a previous PR and are already merged; remember to start new branches off of a clean master rather than a branch for another PR though!

File: lib/DDG/Goodie/ZappBrannigan.pm (hunk 1 above)
Comment (rasikapohankar) [inline lib/DDG/Goodie/ZappBrannigan.pm]: Hey @GuiltyDolphin! Sorry but I didn't get this, what should be cleaned up?

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ZappBrannigan.pm]: You've got trailing white-space (evident on the diff)!

Comment (rasikapohankar) [inline lib/DDG/Goodie/ZappBrannigan.pm]: For the separation of quote selection from the return statement, I think we can keep it?

File: lib/DDG/Goodie/ZappBrannigan.pm [hunk 2]
```diff
@@ -19,12 +19,25 @@ handle query => sub {
     # Ensure rand is seeded for each process
     srand();
     my @quote = @{$quotes->[int(rand(scalar(@$quotes)))]};
-
+    
+    foreach (@quote) {        
+        $_ =~ s/^([^:]+):/$1:<\/i>/;
+        $_ = '<i>'.$_;        
+    }
+
```
Comment (rasikapohankar) [inline lib/DDG/Goodie/ZappBrannigan.pm]: @GuiltyDolphin What do you think of the addition of italics to the names in the quotes? I have tried to do it, the test has failed though. :disappointed:

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ZappBrannigan.pm]: @rasikapohankar I'd leave it as-is; we're only trying to update the structured answer here, not the display.

You could make an issue or a PR in the future if you think italics would look better (and discussion could happen there).

Comment (rasikapohankar) [inline lib/DDG/Goodie/ZappBrannigan.pm]: @GuiltyDolphin Okay, thanks! I'll revert the changes tomorrow. :)

**Comment by rasikapohankar** [comment]
@GuiltyDolphin Removed the code for italics. The test was failing when data was,
       `data => {
            content => qr/Zapp Brannigan: /,
            subtitle => 'Zapp Brannigan quote'            
        }` inside `structured_answer`. It passed on changing data to `data => '-ANY-'`. Why does this happen?

**Comment by GuiltyDolphin** [comment]
@rasikapohankar Don't use `-ANY-` any more - it was only there as a 'hack' to accept *any* data. We can use `Test::Deep` now so much more sophisticated conditions are possible.

File: t/ZappBrannigan.t [hunk 3]
```diff
@@ -13,11 +13,15 @@ ddg_goodie_test(
     'zapp brannigan quote' => test_zci(
         qr/Zapp Brannigan: /,
         structured_answer => {
-            input     => [],
-            operation => 'Zapp Brannigan quote',
-            result    => qr/Zapp Brannigan: /,
+            data => '-ANY-',
```
Comment (GuiltyDolphin) [inline t/ZappBrannigan.t]: Don't use `-ANY-`, we know what data goes through here so we can specify.

**Comment by rasikapohankar** [comment]
@GuiltyDolphin Oh, didn't know this! SunInfo.t seems to have used it, I found it there and tried it. So, data should be kept `data => { content => qr/Zapp Brannigan: /, subtitle => 'Zapp Brannigan quote' }` in the test file? The test fails for this saying,     
   `# Compared reftype($data->{"structured_answer"}{"data"}{"content"})
    # got : undef
    # expect : 'REGEXP'` on line 77 of Goodie.pm. I didn't understand why it is failing.

File: lib/DDG/Goodie/ZappBrannigan.pm (hunk 1 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/ZappBrannigan.pm]: @rasikapohankar I mean you've got spaces ***on their own*** on a line - you can have empty lines.

**Comment by GuiltyDolphin** [comment]
@rasikapohankar Don't use a raw regex in the tests, use `re(...)` from `Test::Deep` (worth reading the Perl documentation for `perldoc Test::Deep`).

**Comment by rasikapohankar** [comment]
@GuiltyDolphin Thanks, it passes now. I realize now, I had read the same discussion on Slack but it hadn't made sense then. So much to learn here! :smile:

**Comment by GuiltyDolphin** [comment]
@rasikapohankar Thanks :dolphin:

**Pull request merged**